File: nodejs/wordwrap/wordwrap.js
exports.SplitTextIntoParagraphs = function(lines) {
    let paragraph = "";
    let paragraphs = [];
    lines.forEach(line => {
        if(line !== "") {
            if(paragraph === "")
                paragraph = line;
            else
                paragraph += " " + line;
        }
        else if(paragraph !== "") {
            paragraphs.push(paragraph);
            paragraph = "";
        }
    });
    if(paragraph !== "") {
        paragraphs.push(paragraph);
    }
    return paragraphs;
};

exports.SplitParagraphsIntoWords = function(paragraphs) {
    let words = [];
    paragraphs.forEach(paragraph => {
        let wordsOfParagraph = [];
        paragraph.split(" ")
            .filter(w => w !== "")
            .forEach(word =>
                wordsOfParagraph.push(word));
        words.push(wordsOfParagraph);
    });
    return words;
};

exports.CreateLines = function(paragraphs, length) {
    let lines = [];
    let line = "";

    paragraphs.forEach(paragraph => {
        if(lines.length > 0) {
            lines.push("");
        }
        paragraph.forEach(word => {
            let newline = line + " " + word;
            if(line === "")
                line = word;
            else if(newline.length <= length)
                line = newline
            else {
                lines.push(line);
                line = word;
            }
        });
        if(line !== "") {
            lines.push(line);
            line = "";
        }
    });

    return lines;
};

File: python/wordwrap/wordwrap.py
import re


def word_wrap(text, max_length):
    paragraphs = _split_text_into_paragraphs(text)
    words_per_paragraph = _split_paragraphs_into_words(paragraphs)
    lines = _create_lines(words_per_paragraph, max_length)
    return lines


def _split_text_into_paragraphs(text):
    text = re.sub(r'\n\n+', '\n\n', text)
    return text.split('\n\n')


def _split_paragraphs_into_words(paragraphs):
    for paragraph in paragraphs:
        yield paragraph.replace('\n', ' ').split(' ')


def _create_lines(words_per_paragraph, max_length):
    for words in words_per_paragraph:
        lines = _create_lines_for_paragraph(words, max_length)
        yield from lines
        yield ''


def _create_lines_for_paragraph(words, max_length):
    line = ''
    for word in words:
        if line == '':
            line = word
        else:
            new_line = line + ' ' + word
            if len(new_line) <= max_length:
                line = new_line
            else:
                yield line
                line = word
    if line != '':
        yield line
